pkg/servicemesh/federation/status/handler.go

```diff
@@ -0,0 +1,448 @@
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	mapping := statusMappingFor(service, exportedName)
+	if existing, ok := h.exports[mapping.LocalService]; !ok {
+		h.logger.Debugf("ExportUpdated called when export mapping does not exist: %+v", mapping.LocalService)
+	} else if reflect.DeepEqual(existing, mapping) {
+		return
+	}
+	h.exports[mapping.LocalService] = mapping
+	h.exportsDirty = true
+}
+
+func (h *handler) ExportRemoved(service model.ServiceKey) {
+	h.logger.Debugf("%s.ExportRemoved(%+v)", h.mesh, service)
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	key := statusServiceKeyFor(service)
+	if _, ok := h.exports[key]; !ok {
+		h.logger.Debugf("ExportRemoved called when export mapping does not exist: %+v", key)
+		return
+	}
+	delete(h.exports, key)
+	h.exportsDirty = true
+}
+
+// Imports
+func (h *handler) ImportAdded(service model.ServiceKey, exportedName string) {
+	h.logger.Debugf("%s.ImportAdded(%+v, %s)", h.mesh, service, exportedName)
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	mapping := statusMappingFor(service, exportedName)
+	if existing, ok := h.imports[mapping.ExportedName]; ok {
+		h.logger.Debugf("ImportAdded called when import mapping already exists: existing=%+v, new=%+v", existing, mapping)
+		if reflect.DeepEqual(existing, mapping) {
+			return
+		}
+	}
+	h.imports[mapping.ExportedName] = mapping
+	h.importsDirty = true
+}
+
+func (h *handler) ImportUpdated(service model.ServiceKey, exportedName string) {
+	h.logger.Debugf("%s.ImportUpdated(%+v, %s)", h.mesh, service, exportedName)
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	mapping := statusMappingFor(service, exportedName)
+	if existing, ok := h.imports[mapping.ExportedName]; !ok {
+		h.logger.Debugf("ImportUpdated called when import mapping does not exist: %s", mapping.ExportedName)
+	} else if reflect.DeepEqual(existing, mapping) {
+		return
+	}
+	h.imports[mapping.ExportedName] = mapping
+	h.importsDirty = true
+}
+
+func (h *handler) ImportRemoved(exportedName string) {
+	h.logger.Debugf("%s.ImportRemoved(%s)", h.mesh, exportedName)
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	if _, ok := h.imports[exportedName]; !ok {
+		h.logger.Debugf("ImportRemoved called when import mapping does not exist: %s", exportedName)
+		return
+	}
+	delete(h.imports, exportedName)
+	h.importsDirty = true
+}
+
+func (h *handler) isDirty() bool {
+	return h.dirty || h.exportsDirty || h.importsDirty || h.discoveryDirty
+}
+
+func (h *handler) pruneOldRemotes() {
+	var connections []string
+	for connection, status := range h.discovery {
+		if status.Connected || metav1.Now().Sub(status.LastDisconnect.Time) < staleRemoteStatusDuration {
+			continue
+		}
+		h.logger.Debugf("removing stale remote watch status for connection from %s", connection)
+		connections = append(connections, connection)
+	}
+	h.discoveryDirty = h.discoveryDirty || len(connections) > 0
+	for _, connection := range connections {
+		delete(h.discovery, connection)
+	}
+}
+
+// Write status
+func (h *handler) Flush() error {
```

Comment:
nit: I think this could use a comment when to use `Flush()` and when it's not required (some funcs seem to call `Flush()` but some don't)